drivers/charger/bq24190.c

```diff
@@ -0,0 +1,485 @@
+/*
+ * Copyright 2023 Cirrus Logic, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT ti_bq24190
+
+#include <errno.h>
+
+#include "bq24190.h"
+
+#include "zephyr/device.h"
+#include "zephyr/drivers/charger.h"
+#include "zephyr/drivers/i2c.h"
+#include "zephyr/kernel.h"
+#include "zephyr/sys/util.h"
+#include "zephyr/logging/log.h"
+
+LOG_MODULE_REGISTER(ti_bq24190);
+
+static int bq24190_register_reset(const struct device *dev)
+{
+	const struct bq24190_config *const config = dev->config;
+	int ret, limit = 100;
+	uint8_t val;
+
+	ret = i2c_reg_update_byte_dt(&config->i2c, BQ24190_REG_POC, BQ24190_REG_POC_RESET_MASK,
+				     BQ24190_REG_POC_RESET_MASK);
+	if (ret) {
+		return ret;
+	}
+
+	do {
+		ret = i2c_reg_read_byte_dt(&config->i2c, BQ24190_REG_POC, &val);
+		if (ret) {
+			return ret;
+		}
+
+		if (!(val & BQ24190_REG_POC_RESET_MASK)) {
+			return 0;
+		}
+
+		k_usleep(100);
+	} while (--limit);
+
+	return -EIO;
+}
+
+static int bq24190_charger_get_charge_type(const struct device *dev, union charger_propval *val)
+{
+	const struct bq24190_config *const config = dev->config;
+	uint8_t v;
+	int ret;
+
+	val->charge_type = CHARGER_CHARGE_TYPE_UNKNOWN;
+
+	ret = i2c_reg_read_byte_dt(&config->i2c, BQ24190_REG_POC, &v);
+	if (ret) {
+		return ret;
+	}
+
+	v = FIELD_GET(BQ24190_REG_POC_CHG_CONFIG_MASK, v);
+
+	if (!v) {
+		val->charge_type = CHARGER_CHARGE_TYPE_NONE;
+	} else {
+		ret = i2c_reg_read_byte_dt(&config->i2c, BQ24190_REG_CCC, &v);
+		if (ret) {
+			return ret;
+		}
+
+		v = FIELD_GET(BQ24190_REG_CCC_FORCE_20PCT_MASK, v);
+
+		if (v) {
+			val->charge_type = CHARGER_CHARGE_TYPE_TRICKLE;
+		} else {
+			val->charge_type = CHARGER_CHARGE_TYPE_FAST;
+		}
+	}
+
+	return 0;
+}
+
+static int bq24190_charger_get_health(const struct device *dev, union charger_propval *val)
+{
+	const struct bq24190_config *const config = dev->config;
+	uint8_t v;
+	int ret;
+
+	ret = i2c_reg_read_byte_dt(&config->i2c, BQ24190_REG_F, &v);
+	if (ret) {
+		return ret;
+	}
+
+	if (v & BQ24190_REG_F_NTC_FAULT_MASK) {
+		switch (v >> BQ24190_REG_F_NTC_FAULT_SHIFT & 0x7) {
+		case 0x1: /* TS1 Cold */
+		case 0x3: /* TS2  Cold */
+		case 0x5: /* Both Cold */
```

Comment:
these should be define rather than comments

Reply:
ack